driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java

```diff
@@ -50,30 +66,89 @@
-        private Simple(Host host) {
+    static class DefaultConvictionPolicy extends ConvictionPolicy {
+        private final Host host;
+        private final ReconnectionPolicy reconnectionPolicy;
+        private final AtomicInteger openConnections = new AtomicInteger();
+
+        private volatile long nextReconnectionTime = Long.MIN_VALUE;
+        private ReconnectionPolicy.ReconnectionSchedule reconnectionSchedule;
+
+        private DefaultConvictionPolicy(Host host, ReconnectionPolicy reconnectionPolicy) {
+            this.host = host;
+            this.reconnectionPolicy = reconnectionPolicy;
+        }
+
+        @Override
+        void signalConnectionCreated() {
+            int newCount = openConnections.incrementAndGet();
+            Host.statesLogger.debug("[{}] new connection created, total = {}", host, newCount);
+            resetReconnectionTime();
+        }
+
+        @Override
+        void signalConnectionClosed() {
+            if (host.state == Host.State.DOWN)
+                return;
+            int remaining = openConnections.decrementAndGet();
+            assert remaining >= 0;
+            Host.statesLogger.debug("[{}] connection closed, remaining = {}", host, remaining);
         }
 
         @Override
-        public boolean addFailure(ConnectionException exception) {
-            return true;
+        boolean signalConnectionFailure(ConnectionException exception, boolean wasFullyInitialized) {
+            if (host.state == Host.State.DOWN)
+                return false;
+
+            updateReconnectionTime();
+            if (wasFullyInitialized) {
+                int remaining = openConnections.decrementAndGet();
+                assert remaining >= 0;
+                Host.statesLogger.debug("[{}] remaining connections = {}", host, remaining);
+                return remaining == 0;
+            } else {
+                return false;
+            }
+        }
+
+        private synchronized void updateReconnectionTime() {
+            long now = System.currentTimeMillis();
+            if (nextReconnectionTime > now)
+                // Someone else updated the time before us
+                return;
+
+            if (reconnectionSchedule == null)
+                reconnectionSchedule = reconnectionPolicy.newSchedule();
+
+            long nextDelayMs = reconnectionSchedule.nextDelayMs();
+            Host.statesLogger.debug("[{}] preventing new connections for the next {} ms", host, nextDelayMs);
+            nextReconnectionTime = now + nextDelayMs;
         }
 
-        public boolean addFailureFromExternalDetector() {
-            return true;
+        private synchronized void resetReconnectionTime() {
+            reconnectionSchedule = null;
+            nextReconnectionTime = Long.MIN_VALUE;
         }
 
         @Override
-        public void reset() {}
+        boolean canReconnectNow() {
+            return System.currentTimeMillis() >= nextReconnectionTime;
```

Comment:
It seems weird to me for the ConvictionPolicy to dictate when to reconnect.  I think it would be much easier to manage if it was only responsible for whether or not to mark the connection down, especially if we ever intend to expose this.